src/iface/route.rs

```diff
@@ -0,0 +1,207 @@
+/// ```
+#[derive(Debug)]
+pub struct Routes<'a> {
+    storage:      ManagedMap<'a, IpCidr, Route>,
+}
+
+impl<'a> Routes<'a> {
+    /// Creates a routing tables. The backing storage is **not** cleared
+    /// upon creation.
+    pub fn new<T>(storage: T) -> Routes<'a>
+            where T: Into<ManagedMap<'a, IpCidr, Route>> {
+        let storage = storage.into();
+        Routes { storage }
+    }
+
+    /// Update the routes of this node.
+    ///
+    /// # Panics
+    /// This function panics if any of the addresses are not unicast.
+    pub fn update_routes<F: FnOnce(&mut ManagedMap<'a, IpCidr, Route>)>(&mut self, f: F) {
+        f(&mut self.storage);
+    }
+
+    /// Add a default ipv4 gateway (ie. "ip route add 0.0.0.0/0 via `gateway`").
+    ///
+    /// On success, returns the previous default gateway, if any.
+    ///
+    /// If not enough storage, returns the gateway address.
+    #[cfg(feature = "proto-ipv4")]
+    pub fn add_default_ipv4_gateway(&mut self, gateway: Ipv4Address) -> Result<Option<Route>, IpAddress> {
+        let cidr = IpCidr::new(IpAddress::v4(0, 0, 0, 0), 0);
+        let route = Route::new_ipv4_gateway(gateway);
+        match self.storage.insert(cidr, route) {
+            Ok(route) => Ok(route),
+            Err((_cidr, route)) => Err(route.via_router)
+        }
+    }
+
+    /// Add a default ipv6 gateway (ie. "ip -6 route add ::/0 via `gateway`").
+    ///
+    /// On success, returns the previous default gateway, if any.
+    ///
+    /// If not enough storage, returns the gateway address.
+    #[cfg(feature = "proto-ipv6")]
+    pub fn add_default_ipv6_gateway(&mut self, gateway: Ipv6Address) -> Result<Option<Route>, IpAddress> {
+        let cidr = IpCidr::new(IpAddress::v6(0, 0, 0, 0, 0, 0, 0, 0), 0);
+        let route = Route::new_ipv6_gateway(gateway);
+        match self.storage.insert(cidr, route) {
+            Ok(route) => Ok(route),
+            Err((_cidr, route)) => Err(route.via_router)
+        }
+    }
+
+    pub(crate) fn lookup(&self, addr: &IpAddress, timestamp: Instant) ->
+            Option<IpAddress> {
+        if addr.is_broadcast() {
+            panic!("IP address {} is broadcast", addr);
+        }
+
+        let cidr = match addr {
+            #[cfg(feature = "proto-ipv4")]
+            IpAddress::Ipv4(addr) => IpCidr::Ipv4(Ipv4Cidr::new(*addr, 32)),
+            #[cfg(feature = "proto-ipv6")]
+            IpAddress::Ipv6(addr) => IpCidr::Ipv6(Ipv6Cidr::new(*addr, 128)),
+            _ => unimplemented!()
+        };
+
+        for (prefix, route) in self.storage.range((Bound::Unbounded, Bound::Included(cidr))).rev() {
+            // TODO: do something with route.preferred_until
+            if let Some(expires_at) = route.expires_at {
+                if timestamp > expires_at {
+                    continue;
+                }
+            }
+
+            if route.via_router != *addr && prefix.contains_addr(addr) {
```

Comment:
I don't think this is necessary. If the address we're querying is the address of the router we either didn't run address resolution or it didn't respond. If `lookup` returns the address of the router we'll either used the cached link layer address or run address resolution.